### ISSUE TITLE
Layer/Regularizer/Loss verification progress

### ISSUE THREAD

**Issue by shijieliu** [NVIDIA/HugeCTR#144]
Description:

| Layer | Equivalent TensorFlow Code | Status |
| --- | --- | --- |
| Concat | `tf.concat` | PASS |
| Dropout | `tf.nn.dropout` | PASS<br>~~Scaling~~!16 |
| Elu | `tf.nn.elu` | PASS |
| FmOrder2 | `def fm_order2(x):`<br>`  x_sum = tf.reduce_sum(x, 1)`<br>`  x_sum_squared = tf.square(x_sum)`<br>`  x_squared = tf.square(x)`<br>`  x_squared_sum = tf.reduce_sum(x_squared, 1)`<br>`return (x_sum_squared - x_squared_sum) / 2` | Precision Issue |
| MultiCross | `def multi_cross(x, w, b, layers):`<br>`  y = []`<br>`  for i in range(layers):`<br>`    v = tf.linalg.matvec(x if i == 0 else y[i - 1], w[i])`<br>`    v = tf.transpose(v)`<br>`    m = tf.multiply(x, v)`<br>`    m = tf.add(m, x if i == 0 else y[i - 1])`<br>`    m = tf.add(m, b[i])`<br>`    y.append(m)`<br>`  return y` | ~~Backward Result Inconsistent (layers > 1)~~ !19<br>Precision Issue |
| Multiply | `tf.math.multiply` | Precision Issue<br>~~Atomic~~!18 |
| Relu | `tf.nn.relu` | PASS |
| Slice | `def slice(x, offsets, lengths):`<br>`  y = []`<br>`  for i in zip(offsets, lengths):`<br>`    y.append(tf.slice(x, [0, i[0]], [-1, i[1]]))`<br>`  return y` | PASS |
| Reshape | `def reshape(x, indices):`<br>`  if indices:`<br>`    x = tf.gather(x, indices, axis=1)`<br>`  return tf.reshape(x, [x.shape[0], -1])` | PASS |
| Add |`tf.add_n(xs)` | PASS |
| FullyConnected |`def tf_fully_connected(x, w, bias):`<br>`    return tf.matmul(x, w) + bias` | PASS |
| BatchNorm | `def tf_batch_norm(x, gamma, beta, momentum, eps):`<br>`    moving_mean = tf.zeros(shape=(1, feature_dim))`<br>`    moving_var = tf.zeros(shape=(1, feature_dim))`<br>`    mean_var = tf.nn.moments(x, axes=0, keepdims=True)`<br>`    y = tf.nn.batch_normalization(x, mean_var[0], mean_var[1], beta, gamma, eps)`<br>`    moving_mean = momentum * moving_mean + (1 - momentum) * mean_var[0]`<br>`     moving_var = momentum * moving_var + (1 - momentum) * (mean_var[1] * batch_size / (batch_size - 1)) `<br>`    return y, moving_mean, moving_var`| Partial PASS<br>gamm_grad inconsistent|
| Embedding | `def tf_embedding(input_keys, embedding_table):`<br>`    """`<br>`    input_keys: indexes: [batchsize, slot_num, max_nnz_per_slot]`<br>`        -1: represent placeholder number, `<br>`        mapping it to the last index whose values are 0.`<br>`    """`<br>`    vocabulary_size, embedding_vec_size = embedding_table.shape`<br>`    input_keys = np.array(input_keys)`<br>`    input_keys[input_keys == -1] = vocabulary_size`<br>`    zeros = np.zeros(shape=(1, embedding_vec_size))`<br>`    embedding_table = np.concatenate((embedding_table, zeros), axis=0)`<br>`    emb = tf.keras.layers.Embedding(vocabulary_size + 1, embedding_vec_size,`<br>`            embeddings_initializer = tf.constant_initializer(value=embedding_table))`<br>`    out = emb(input_keys)`<br>`    out = tf.reduce_sum(out, axis=-2)`<br>`    return out`| PASS |

| Regularizer | Equivalent TensorFlow Code | Status |
| --- | --- | --- |
| L1Regularizer |`def tf_l1regularizer(weights, lamda, batchsize):`<br>`    tf_l1 = tf.keras.regularizers.l1(l=lamda)`<br>`    tf_rterm = tf_l1(weights)`<br>`    tf_rterm /= batchsize`<br>`    return tf_rterm` | Partial PASS<br> grad inconsistent when weight=0 <br><img src="/uploads/bce4f717128b9a4f2815858f7f938a2e/grad_pic.png" width=200/> |
| L2Regularizer |`def tf_l2regularizer(weights, lamda, batchsize):`<br>`    tf_l2 = tf.keras.regularizers.l2(l=lamda)`<br>`    tf_rterm = tf_l2(weights)`<br>`    tf_rterm /= (batchsize * 2)`<br>`    return tf_rterm` | PASS |

| Loss | Equivalent TensorFlow Code | Status |
| --- | --- | --- |
| BinaryCrossEntropy | `bce = tf.keras.losses.BinaryCrossentropy(from_logits=True)`<br>`loss = bce(y, x)`| PASS |
| CrossEntropyLoss | `cel = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)`<br>`loss = cel(y, x)`| PASS | 
| MultiCrossEntropy | `def tf_multi_cross_entropy(x, y, target_weights):`<br>`    bce = tf.keras.losses.BinaryCrossentropy(from_logits=True)`<br>`    totalLoss = 0`<br>`    for i in range(batch_size):`<br>`        x_ = tf.reshape(x[i], x[i].shape + (1,))`<br>`        y_ = tf.reshape(y[i], y[i].shape + (1,))`<br>`        loss = bce(y_, x_, sample_weight = target_weights)`<br>`        totalLoss += loss`<br>`    totalLoss /= (batch_size)`<br>`    return totalLoss`| ~~Precision Issue<br>Due to exponential operation~~!44 |

| Optimizers in Embedding | Equivalent TensorFlow Code | Status |
| --- | --- | --- |
| Adam | `tf.keras.optimizers.Adam()` | Consistent using **Global** Update|
| MomentumSGD | `tf.keras.optimizers.SGD(momentum=momentum_rate)` | Consistent using **Local** Update |
| Nesterov    | `tf.keras.optimizers.SGD(momentum=momentum_rate, `<br>`nesterov=True)` | Consistent Using **Local** Update |

###  **Precision Issue** :

Precision issue is a typical problem, which exists in many layers.
The main reason is that use the serial reduction approach when accumulating.

Refer to the comparison for different approach below:

**array length is 1024**

Approach|Absolute difference sum(1000 times)|Relative difference sum(1000 times)
---|---|---
1024bit float serial accumulation|0.000000|0.000000
64bit float serial accumulation|0.001089|0.000042
32bit float serial accumulation|0.013122|0.001410
32bit binary merge accumulation|0.002341|0.000248

**array length is 10000**

Approach|Absolute difference sum(1000 times)|Relative difference sum(1000 times)
---|---|---
1024bit float serial accumulation|0.000000|0.000000
64bit float serial accumulation|0.003819|0.000043
32bit float serial accumulation|0.127352|0.004792
32bit binary merge accumulation|0.008906|0.000356

Use binary merge reduce can get obvious accuracy compare to serial accumulation, for details can refer to [here](https://goooxu.github.io/blog/cuda_programming/precision_problem.html)

### **`MultiCrossLayer` Backward Result Inconsistent** (**FIXED**):

The backward result of `MultiCrossLayer` is inconsistent to TensorFlow's, if layers exceed 1.
But the utest of MultiCrossLayer has passed with same parameters.

## Build & Test ###
Checkout the test branch via
```
$ git checkout fea-tensorflow-op
```
Build via
```
$ mkdir -p build
$ cd build
$ cmake -DCMAKE_BUILD_TYPE=Release -DSM=XX ..
$ make
$ cd ..
```
then run the script via
```
$ python3 tf_op/python/ops/test.py
```
Comments:
